## Воркшоп #3: Робота з Pandas, обʼєднання і агрегація даних

Цей воркшоп виконує...

Почнемо! 
Перш за все, зчитайте 4 набори даних: `courses.csv`, `enrollment.csv`, `instructors.csv` та `students_v2.csv` та запишіть їх у змінні `courses`, `enrollment`, `instructors`, `students` відповідно. Обовʼязково перевірте, що дані відповідають очікуванням.

In [35]:
import pandas as pd

In [36]:
courses = pd.read_csv("courses.csv", delimiter= ",")
enrollment = pd.read_csv("enrollment.csv", delimiter=",")
instructors = pd.read_csv("instructors.csv", delimiter=",")
students = pd.read_csv("students_v2.csv", delimiter=",")

In [44]:
print(students.columns)

Index(['Student_ID', 'Name', 'Age', 'Faculty', 'Major', 'GPA', 'Start_Date',
       'First_semester'],
      dtype='object')


In [45]:
print(courses.columns)

Index(['Course_ID', 'Course_Name', 'Faculty', 'Credits', 'Instructor'], dtype='object')


In [46]:
print(enrollment.columns)

Index(['Enrollment_ID', 'Student_ID', 'Course_ID', 'First_semester', 'Grade'], dtype='object')


In [45]:
instructors

Instructor_ID                Name      Faculty  \
0          TDDUH   Pamela Farmer DDS     Business   
1          AE9SI        Brian Holden      Science   
2          FYRP6       Michael Patel      Science   
3          X5GWV     Brittany Mathis  Engineering   
4          J8MV0  Christopher Cooper     Medicine   
5          F2020    Jamie Jordan Jr.     Medicine   
6          40MYX         Glenn Smith          Law   
7          6C4MF          Mary Scott  Engineering   
8          JB603         Diana Greer  Engineering   
9          CQ1XF        Lauren David          Law   
10         I3AH7     Jeffrey Collins          Law   
11         X9BIG         Diana Scott  Engineering   
12         31OFB    Michael Reynolds         Arts   
13         0LO2L        Rachel Miles     Business   
14         1R89R       Andrew Gaines     Medicine   
15         VNJEO       Stephanie Cox     Medicine   
16         2K4JD        Henry Graves      Science   
17         8M3RA          Dana Lopez     Medicine   
18         2A3U7      Cynthia Reeves         Arts   
19         NFYMU        Ashlee Lopez     Medicine   
20         0AZDV        James Nelson     Business   
21         GW4P4      Garrett Parker         Arts   
22         P38QS        James Graham      Science   
23         MLUFZ    Samantha Aguilar  Engineering   
24         AH7AE         Diana Short         Arts   
25         43JQ0        Brad Hancock         Arts   
26         82EU8          Megan Ruiz          Law   
27         R83JF        Allison Long      Science   
28         OFSI9     Jennifer Thomas      Science   
29         UO5C3        Carla Orozco     Business   
30         MV9DL         Misty Blair     Business   
31         4L36O      Melissa Hughes     Business   
32         SHHID     Angela Thompson         Arts   
33         BHNRN       Michael Casey  Engineering   
34         VML71         Nicole Cole         Arts   
35         IL1XD          Jose Huang     Business   
36         JV74G        Richard Carr         Arts   
37         O43EQ        Susan Taylor         Arts   
38         SK7O2        Joseph Smith      Science   
39         IYJ5P        Laura Rogers      Science   
40         ZECD3      Mallory Taylor  Engineering   
41         DI9PP       George Lawson          Law   
42         RQBWY        Rebecca Buck          Law   
43         2TV3J     Thomas Mcintosh          Law   
44         HOFLI          Lisa Scott     Medicine   
45         JWGTU  Christopher Chavez     Business   
46         ZZS8A       Brent Johnson     Medicine   
47         KUND4    Stephanie Morris          Law   
48         YCJMK         Karen Hodge     Medicine   
49         1Z96C       Tammy Johnson      Science   

                             Email  
0    pamela.farmer.dds@univer.sity  
1         brian.holden@univer.sity  
2        michael.patel@univer.sity  
3      brittany.mathis@univer.sity  
4   christopher.cooper@univer.sity  
5     jamie.jordan.jr.@univer.sity  
6          glenn.smith@univer.sity  
7           mary.scott@univer.sity  
8          diana.greer@univer.sity  
9         lauren.david@univer.sity  
10     jeffrey.collins@univer.sity  
11         diana.scott@univer.sity  
12    michael.reynolds@univer.sity  
13        rachel.miles@univer.sity  
14       andrew.gaines@univer.sity  
15       stephanie.cox@univer.sity  
16        henry.graves@univer.sity  
17          dana.lopez@univer.sity  
18      cynthia.reeves@univer.sity  
19        ashlee.lopez@univer.sity  
20        james.nelson@univer.sity  
21      garrett.parker@univer.sity  
22        james.graham@univer.sity  
23    samantha.aguilar@univer.sity  
24         diana.short@univer.sity  
25        brad.hancock@univer.sity  
26          megan.ruiz@univer.sity  
27        allison.long@univer.sity  
28     jennifer.thomas@univer.sity  
29        carla.orozco@univer.sity  
30         misty.blair@univer.sity  
31      melissa.hughes@univer.sity  
32     angela.thompson@univer.sity  
33       michael.casey@univer.sity  
34  

Тепер, коли ми маємо усі дані на руках, давайте спробуємо їх обʼєднати. Використайте `pd.merge`, щоб обʼєднати студентів з тими курсами, які вони брали.

Обовʼязково зверніть увагу на ключі, за якими ви обʼєднуєте дані! "Природнє" зʼєднання тут не прокатить, здогадаєтесь, чому? 
 

In [37]:
df = pd.merge(students, enrollment, on="Student_ID", how="left")
df = pd.merge(df, courses, on="Course_ID", how="left")


In [38]:
df

Student_ID             Name  Age Faculty_x          Major   GPA  \
0        UEPE0     Kelly Greene   19       Law  Environmental  2.91   
1        C15ED    Joseph George   20  Medicine        General  2.79   
2        C15ED    Joseph George   20  Medicine        General  2.79   
3        C15ED    Joseph George   20  Medicine        General  2.79   
4        QVQDT  Cynthia Hancock   23  Business        Finance  2.24   
..         ...              ...  ...       ...            ...   ...   
206      QSVRU   Donna Harrison   22  Medicine     Psychiatry  2.93   
207      9O3RS   Michael Bryant   22      Arts     Philosophy  2.33   
208      9O3RS   Michael Bryant   22      Arts     Philosophy  2.33   
209      E6STQ  Patrick Johnson   18  Business        Finance  2.97   
210      E6STQ  Patrick Johnson   18  Business        Finance  2.97   

     Start_Date First_semester_x Enrollment_ID Course_ID First_semester_y  \
0    2020-01-20      Spring 2020         1KEUI     1EKBM        Fall 2021   
1    2023-07-04      Summer 2023         BYB6X     1GBT1        Fall 2020   
2    2023-07-04      Summer 2023         9ZSI3     ZTEEQ        Fall 2021   
3    2023-07-04      Summer 2023         L5Q5V     HNU9L      Spring 2023   
4    2019-07-30      Summer 2019         0JI0A     JLM1K      Spring 2023   
..          ...              ...           ...       ...              ...   
206  2019-09-13        Fall 2019           NaN       NaN              NaN   
207  2020-06-03      Summer 2020         AF8IX     E5TPT        Fall 2021   
208  2020-06-03      Summer 2020         H17KY     TC1IL        Fall 2021   
209  2024-05-09      Spring 2024         51X9B     ZTEEQ        Fall 2021   
210  2024-05-09      Spring 2024         VJNG4     0BF6M        Fall 2021   

    Grade         Course_Name    Faculty_y  Credits       Instructor  
0       F       World History         Arts      4.0  Angela Thompson  
1       A  Constitutional Law          Law      5.0  Jeffrey Collins  
2       C             Anatomy     Medicine      4.0       Lisa Scott  
3       F     Quantum Physics      Science      5.0     Henry Graves  
4       A             Statics  Engineering      4.0       Mary Scott  
..    ...                 ...          ...      ...              ...  
206   NaN                 NaN          NaN      NaN              NaN  
207     B        Property Law          Law      3.0       Megan Ruiz  
208     F        Neuroscience     Medicine      4.0    Andrew Gaines  
209     A             Anatomy     Medicine      4.0       Lisa Scott  
210     D        Pharmacology     Medicine      3.0      Karen Hodge  

[211 rows x 16 columns]

Якщо ви все зробили правильно, у вас мало вийти 200 записів. Давайте сконцетруємось на записах студента на імʼя `Gregory Martin` (або будь-якого іншого/іншої, який/яка вам до вподоби) та подивимось, які саме курси він взяв. Для цього вам знадобляться вже три датасети. 


In [39]:
gmartin_courses = df[df["Name"] == "Gregory Martin"]
inf_gmartin_courses = gmartin_courses[["Name", "Major", "Course_Name", "Grade", "Instructor"]]
inf_gmartin_courses

Name    Major           Course_Name Grade     Instructor
176  Gregory Martin  Physics              Calculus     D   Brian Holden
177  Gregory Martin  Physics              Calculus     A  Tammy Johnson
178  Gregory Martin  Physics            Modern Art     D    Nicole Cole
179  Gregory Martin  Physics              Circuits     A    Diana Greer
180  Gregory Martin  Physics  Financial Accounting     A   Rachel Miles

І, ніби нам цього було мало, дізнаємось імейли інструкторів, які ведуть курси, на які Грегорі був записаний:

In [44]:
df = pd.merge(df, instructors, on="Faculty", how="left")
df

Student_ID           Name_x  Age Faculty_x          Major   GPA  \
0        UEPE0     Kelly Greene   19       Law  Environmental  2.91   
1        C15ED    Joseph George   20  Medicine        General  2.79   
2        C15ED    Joseph George   20  Medicine        General  2.79   
3        C15ED    Joseph George   20  Medicine        General  2.79   
4        QVQDT  Cynthia Hancock   23  Business        Finance  2.24   
..         ...              ...  ...       ...            ...   ...   
206      QSVRU   Donna Harrison   22  Medicine     Psychiatry  2.93   
207      9O3RS   Michael Bryant   22      Arts     Philosophy  2.33   
208      9O3RS   Michael Bryant   22      Arts     Philosophy  2.33   
209      E6STQ  Patrick Johnson   18  Business        Finance  2.97   
210      E6STQ  Patrick Johnson   18  Business        Finance  2.97   

     Start_Date First_semester_x Enrollment_ID Course_ID  ...  \
0    2020-01-20      Spring 2020         1KEUI     1EKBM  ...   
1    2023-07-04      Summer 2023         BYB6X     1GBT1  ...   
2    2023-07-04      Summer 2023         9ZSI3     ZTEEQ  ...   
3    2023-07-04      Summer 2023         L5Q5V     HNU9L  ...   
4    2019-07-30      Summer 2019         0JI0A     JLM1K  ...   
..          ...              ...           ...       ...  ...   
206  2019-09-13        Fall 2019           NaN       NaN  ...   
207  2020-06-03      Summer 2020         AF8IX     E5TPT  ...   
208  2020-06-03      Summer 2020         H17KY     TC1IL  ...   
209  2024-05-09      Spring 2024         51X9B     ZTEEQ  ...   
210  2024-05-09      Spring 2024         VJNG4     0BF6M  ...   

            Course_Name    Faculty_y Credits       Instructor  \
0         World History         Arts     4.0  Angela Thompson   
1    Constitutional Law          Law     5.0  Jeffrey Collins   
2               Anatomy     Medicine     4.0       Lisa Scott   
3       Quantum Physics      Science     5.0     Henry Graves   
4               Statics  Engineering     4.0       Mary Scott   
..                  ...          ...     ...              ...   
206                 NaN          NaN     NaN              NaN   
207        Property Law          Law     3.0       Megan Ruiz   
208        Neuroscience     Medicine     4.0    Andrew Gaines   
209             Anatomy     Medicine     4.0       Lisa Scott   
210        Pharmacology     Medicine     3.0      Karen Hodge   

     Instructor_ID_x Faculty Email_x Instructor_ID_y Name_y Email_y  
0                NaN     NaN     NaN             NaN    NaN     NaN  
1                NaN     NaN     NaN             NaN    NaN     NaN  
2                NaN     NaN     NaN             NaN    NaN     NaN  
3                NaN     NaN     NaN             NaN    NaN     NaN  
4                NaN     NaN     NaN             NaN    NaN     NaN  
..               ...     ...     ...             ...    ...     ...  
206              NaN     NaN     NaN             NaN    NaN     NaN  
207              NaN     NaN     NaN             NaN    NaN     NaN  
208              NaN     NaN     NaN             NaN    NaN     NaN  
209              NaN     NaN     NaN             NaN    NaN     NaN  
210              NaN     NaN     NaN             NaN    NaN     NaN  

[211 rows x 22 columns]

### Кінець підготовчої частини

--------------

### Початок спільної частини 

Чудово! 

До цього ми працювали з конкретним студентом, але тепер давайте подивимось на "big picture" і спробуємо проаналізувати дещо. 

Отож, наше наступне завдання: знайдіть кількість невідповідностей у таблицях students і enrollments. Під "невідповідністю" маємо на увазі наявність запису в одному датасеті, але відсутність в іншому. Тут є кілька способів це зробити, спробуйте хоча б два достатньо різних. 

І ось вам підказки, що можна використати в одному з способів:
- параметри `how` та `indicator` у `pd.merge`
- методи `value_counts()` та `query` у датафрейму 


Альтернативий спосіб натомість спробуйте зробити максимально прямолійнійним, так, якби ви не знали про наявлість `indicator`, `value_counts()`, `query` тощо 


In [ ]:
# Спосіб №1 
enrollment.merge(students, how="outer", on="Student_ID", indicator=True)["_merge"].value_counts()

_merge
both          200
right_only     11
left_only       0
Name: count, dtype: int64

In [ ]:
# Спосіб №2

Щоб закінчити епопею з функцією `merge`, давайте спробуємо порахувати кількість різних **пар** колег з одного факультету. При цьому врахуйте, що "я сам собі не колега", а також кожну пару потрібно рахувати рівно один раз.

In [ ]:
instructors_pairs = instructors.merge(instructors, how="cross")
instructors_pairs[(instructors_pairs["Faculty_x"] == instructors_pairs["Faculty_y"]) & (instructors_pairs["Name_x"] != instructors_pairs["Name_y"])
]["Faculty_x"].count()/2

np.float64(185.0)

Має вийти 185, але важливо не "скільки", а "як" :) 

Тепер давайте спробуємо працювати з групуванням і агрегацією. Перш за все, давайте розглянемо групи по окремості.

Для цього вам знадобиться метод `groupby`. Згрупуємо інструкторів за факультетом і виведемо кількість інструкторів для кожного з факультетів, використовуючи цикл: 

```python

for faculty, members in ???
    print(faculty, len(members))

```
(замість ??? потрібно написати правильний вираз) 


In [ ]:
for faculty, members in instructors.groupby(["Faculty"]):
    print(faculty, members["Name"])

('Arts',) 12    Michael Reynolds
18      Cynthia Reeves
21      Garrett Parker
24         Diana Short
25        Brad Hancock
32     Angela Thompson
34         Nicole Cole
36        Richard Carr
37        Susan Taylor
Name: Name, dtype: object
('Business',) 0      Pamela Farmer DDS
13          Rachel Miles
20          James Nelson
29          Carla Orozco
30           Misty Blair
31        Melissa Hughes
35            Jose Huang
45    Christopher Chavez
Name: Name, dtype: object
('Engineering',) 3      Brittany Mathis
7           Mary Scott
8          Diana Greer
11         Diana Scott
23    Samantha Aguilar
33       Michael Casey
40      Mallory Taylor
Name: Name, dtype: object
('Law',) 6          Glenn Smith
9         Lauren David
10     Jeffrey Collins
26          Megan Ruiz
41       George Lawson
42        Rebecca Buck
43     Thomas Mcintosh
47    Stephanie Morris
Name: Name, dtype: object
('Medicine',) 4     Christopher Cooper
5       Jamie Jordan Jr.
14         Andrew Gaines
15   

Ви також можете вивести список інструкторів замість їх кількості, якщо хочете :) 

Тепер давайте спробуємо погрупувати студентів за факультетом і порахувати середню оцінку та середній вік по кожному з факультетів. Для цього в результата groupby треба виділити конкретний набір колонок - у нашому випадку `Age` та `GPA`. Звісно, що для підрахунку середнього вам підійде функція `mean()`

In [ ]:
students.groupby()

students.groupby[]

Якщо ви все зробили правильно, у вас мав вийти якийсь датасет. Спробуйте подивитися, які у цього датасета колонки та який індекс.

Хм, середні значення то звісно добре, але іноді цікавить дещо інше. Спробуйте залишити середнє значення `GPA`, а для `Age` додати також обрахунок мінімального і максимального значення. 

Тепер спробуйте згрупувати не по одній колонці, а по кількох - наприклад, `Faculty` та `Age`. Так само перевірте назви колонок та індекс і подивіться, як це повʼязано з аргументом `groupby`.

In [62]:
ds = students.groupby([ "Age", "Faculty"])["GPA"].mean()
ds.index

MultiIndex([(18,        'Arts'),
            (18,    'Business'),
            (18, 'Engineering'),
            (18,         'Law'),
            (18,    'Medicine'),
            (18,     'Science'),
            (19,        'Arts'),
            (19, 'Engineering'),
            (19,         'Law'),
            (19,    'Medicine'),
            (19,     'Science'),
            (20,        'Arts'),
            (20,    'Business'),
            (20, 'Engineering'),
            (20,    'Medicine'),
            (20,     'Science'),
            (21,    'Business'),
            (21, 'Engineering'),
            (21,         'Law'),
            (21,    'Medicine'),
            (21,     'Science'),
            (22,        'Arts'),
            (22,    'Business'),
            (22, 'Engineering'),
            (22,    'Medicine'),
            (22,     'Science'),
            (23,        'Arts'),
            (23,    'Business'),
            (23, 'Engineering'),
            (23,         'Law'),
          

Щож, сподіваюсь, ± зрозуміло, чому це працює саме так. 

Student_ID              Name  Age   Faculty        Major   GPA  \
0        0I46C       Tanya Lopez   20  Medicine      Surgery  3.23   
1        0LD46       Brad Garcia   21       Law     Criminal  2.37   
2        2DBFW   William Burgess   22  Medicine      General  2.23   
3        2DBFW   William Burgess   22  Medicine      General  2.23   
4        2KITZ  Jennifer Osborne   19       Law    Corporate  2.12   
..         ...               ...  ...       ...          ...   ...   
206      ZQGR3  Morgan Hernandez   19   Science  Mathematics  3.20   
207      ZQGR3  Morgan Hernandez   19   Science  Mathematics  3.20   
208      ZZZ8L      Beverly Levy   21  Medicine      Surgery  3.85   
209      ZZZ8L      Beverly Levy   21  Medicine      Surgery  3.85   
210      ZZZ8L      Beverly Levy   21  Medicine      Surgery  3.85   

     Start_Date First_semester_x Enrollment_ID Course_ID First_semester_y  \
0    2022-12-28        Fall 2022           NaN       NaN              NaN   
1    2020-09-16        Fall 2020         AVZCR     WBI74        Fall 2022   
2    2022-08-28      Summer 2022         E06O2     WBWPP        Fall 2020   
3    2022-08-28      Summer 2022         GWU63     1XYJS        Fall 2021   
4    2021-10-09        Fall 2021         IXBFQ     063S9        Fall 2020   
..          ...              ...           ...       ...              ...   
206  2022-03-18      Spring 2022         T80MR     E89S8        Fall 2020   
207  2022-03-18      Spring 2022         2R4JU     M24VK      Spring 2023   
208  2022-06-26      Summer 2022         CZJTU     N18K0        Fall 2022   
209  2022-06-26      Summer 2022         OTFI7     JZ4IG      Spring 2022   
210  2022-06-26      Summer 2022         6GU3A     M24VK        Fall 2020   

    Grade  
0     NaN  
1       A  
2       C  
3       D  
4       B  
..    ...  
206     B  
207     D  
208     F  
209     C  
210     F  

[211 rows x 12 columns]

Рухаємось далі. Тут вже треба і `merge`, і `groupby`. Для кожного студента з GPA більшим за 3.3 знайдіть кількість курсів, на які студент заенролився і завалив (тобто здав на `D`). Результат посортуйте по кількості таких курсів. Подивимось, скільки відмінників не такі вже й відмінники 😈

Тут у вас, здається, має вийти 10.

In [88]:
students_A = [students["GPA"] > 3.3]
enrollment_students = enrollment.merge(students, on="Student_ID", how="left")
enrollment_students_D = enrollment_students[enrollment_students["Grade"] == "D"]
enrollment_students_D.groupby(["Name"])["Course_ID"].count().sort_values()

Name
Abigail Williams     1
Chad Mills           1
Carl Pitts DVM       1
Brittany Johnson     1
Crystal Thomas       1
Jennifer Osborne     1
James Frank          1
Ian Johnson          1
Katherine Padilla    1
Katherine Mullen     1
Kyle Anderson        1
Jonathon Scott       1
Mark Nelson          1
Marc Johnson         1
Lisa Mccormick       1
Lance King           1
Shelby Cox           1
Raymond Navarro      1
Richard Price        1
Samantha Floyd       1
Melanie Patrick      1
Morgan Hernandez     1
Patrick Johnson      1
Peggy Hale           1
Teresa Martinez      1
Steven Henderson     1
Tracy Frye           1
Troy Jackson         1
William Burgess      1
Brandon Carter       2
Gregory Martin       2
Andrew Villa         2
Mary Wong            2
Adam Robinson        2
Name: Course_ID, dtype: int64

Зверніть увагу, що студентів-відмінників має бути 29, а от завалені курси є лише у частини з них. Якщо у вас результат складається з менш ніж 29 записів, подумайте, як то виправити, щоб були присутні і студенти без "D-шок".

In [90]:
students_A = students[students["GPA"] > 3.3]
enrollmwent_students = enrollment.merge(students_A, on="Student_ID", how="left")
enrollment_students['Failed'] = enrollment_students["Grade"] == "D"
enrollment_students.groupby(['Name'])["Failed"].sum()

Name
Abigail Williams     1
Adam Robinson        2
Alison Nelson        0
Andrew Villa         2
Anna Jimenez         0
                    ..
Tracy Frye           1
Tracy Sanders        0
Troy Jackson         1
Veronica Mcdaniel    0
William Burgess      1
Name: Failed, Length: 89, dtype: int64

Хм, а як загалом картина на факультетах? 
Спробуйте наступне:
- для кожного факультету порахуйте частку оцінок `F`
- для факультетів з більш ніж 20% F-ок виведіть:
  - кількість виставлених оцінок загалом
  - кількість F-ок
  - частку F-ок
  - кількість студентів, які проходили курси
  - кількість студентів, які отримали F-ки

Обовʼязково переконайтесь, що ви не рахуєте студентів двічі.

В рамках цього завдання можете спробувати використати методи `assign` та `filter`, ну і вже знайомі вам `agg` та `query`. 

Нарешті, давайте збережемо дані. Візьміть датасет студентів і скористайтесь методом `to_csv`, щоб зберегти його у файл `other_students.csv`. Відкрийте файл і зверніть увагу на те, чи все "окей".

Імовірно, у вас буде "зайва" колонка. Щоб зрозуміти, що це конкретно за колонка, спробуйте також зберегти результат команди `students.groupby(['Faculty', 'Age'])[['Age', 'GPA']].mean()` у інший файл. 

На цьому етапі у вас має закрастися підозра, що ж це за колонка. Так-так, це була колонка-індекс. Щоб зберегти без неї, передайте у `to_csv` аргумент `index=False` 

### Кінець спільної частини
-----------------
### Початок самостійної (домашньої) частини


##### Завдання №1

Для кожного інструктора розрахуйте:
- Кількість курсів, які він/вона викладають
- Загальну кількість кредитів відповідних курсів
- Загальну кількість студентів у курсах, які веде викладач
- Кількість унікальних студентів (тобто скільки різних облич бачить викладач на своїх парах)
- Статистику по оцінках - як часто (у %) викладач ставить A, B, C, ..., F

Відсортуйте результат за кількістю унікальних студентів. 

Завдання №2

Для кожного студента порахуйте:
- Сумарну кількість кредитів 
- Кількість днів навчання (до сьогоднішньої дати) [1]
- Середню успішність
- Середню успішність, якщо рахувати, що A - 100, B - 90, C - 75, D - 60, F - 0  
- Кількість непрофільних предметів (тобто предметів, які студент взяв з факультету, відмінного від свого)



[1] Зверніть увагу, що ми з вами не працювали раніше з датами. Пропонується спробувати розібратися самостійно - там нічого складного! Вам знадобиться [to_datetime](https://pandas.pydata.org/docs/reference/api/pandas.to_datetime.html), [Series.dt](https://pandas.pydata.org/docs/reference/api/pandas.Series.dt.html#pandas-series-dt), а також стандартний пайтонівський модуль datetime (`from datetime import date` -> `date.today()`).
